File: BookVerse/Market.Frontend/rs1-frontend-2025-26/src/app/modules/admin/catalogs/book/services/isbn-validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

export function isbnValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const value = control.value;
    if (value === null || value === undefined || value === '') return null;

    const clean = String(value).replace(/[\s-]/g, '').toUpperCase();

    if (clean.length === 10 && isValidIsbn10(clean)) return null;
    if (clean.length === 13 && isValidIsbn13(clean)) return null;

    return { isbn: true };
  };
}

function isValidIsbn10(isbn: string): boolean {
  for (let i = 0; i < 9; i++) {
    if (!/[0-9]/.test(isbn[i])) return false;
  }
  const last = isbn[9];
  if (last !== 'X' && !/[0-9]/.test(last)) return false;

  let sum = 0;
  for (let i = 0; i < 9; i++) {
    sum += parseInt(isbn[i], 10) * (10 - i);
  }
  sum += last === 'X' ? 10 : parseInt(last, 10);

  return sum % 11 === 0;
}

function isValidIsbn13(isbn: string): boolean {
  if (!/^\d{13}$/.test(isbn)) return false;

  let sum = 0;
  for (let i = 0; i < 13; i++) {
    const weight = i % 2 === 0 ? 1 : 3;
    sum += parseInt(isbn[i], 10) * weight;
  }
  return sum % 10 === 0;
}

File: BookVerse/backend/BookVerse.Application/Common/IsbnValidator.cs
namespace BookVerse.Application.Common;

public static class IsbnValidator
{
    public static bool IsValid(string? isbn)
    {
        if (string.IsNullOrWhiteSpace(isbn)) return false;

        var clean = new string(isbn.Where(c => !char.IsWhiteSpace(c) && c != '-').ToArray()).ToUpperInvariant();

        return clean.Length switch
        {
            10 => IsValidIsbn10(clean),
            13 => IsValidIsbn13(clean),
            _ => false
        };
    }

    private static bool IsValidIsbn10(string isbn)
    {
        for (int i = 0; i < 9; i++)
            if (!char.IsDigit(isbn[i])) return false;

        var last = isbn[9];
        if (last != 'X' && !char.IsDigit(last)) return false;

        int sum = 0;
        for (int i = 0; i < 9; i++)
            sum += (isbn[i] - '0') * (10 - i);
        sum += last == 'X' ? 10 : (last - '0');

        return sum % 11 == 0;
    }

    private static bool IsValidIsbn13(string isbn)
    {
        for (int i = 0; i < 13; i++)
            if (!char.IsDigit(isbn[i])) return false;

        int sum = 0;
        for (int i = 0; i < 13; i++)
        {
            int weight = i % 2 == 0 ? 1 : 3;
            sum += (isbn[i] - '0') * weight;
        }
        return sum % 10 == 0;
    }
}
